### PR TITLE
refactor(SwitchNetwork): revamp RPC `starkNet_switchNetwork`

### PR DESCRIPTION
**Pull request by stanleyyconsensys**
This PR is to revamp the RPC `starkNet_switchNetwork` handler to align with new RPC standard

### Requirement:
Requires this PR to be merged first:

- [x] #369
- [x]  #373 
- [x]  #374

### CODE REVIEW

File: packages/starknet-snap/src/rpcs/switch-network.test.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import type { constants } from 'starknet';
+
+import { Config } from '../config';
+import { NetworkStateManager } from '../state/network-state-manager';
+import type { Network } from '../types/snapState';
+import {
+  STARKNET_SEPOLIA_TESTNET_NETWORK,
+  STARKNET_MAINNET_NETWORK,
+} from '../utils/constants';
+import { prepareConfirmDialog } from './__tests__/helper';
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.test.ts:14]:
```suggestion
import {
  UserRejectedOpError,
  InvalidNetworkError,
  InvalidRequestParamsError,
} from '../utils/exceptions';
import { prepareConfirmDialog } from './__tests__/helper';
```

File: packages/starknet-snap/src/rpcs/switch-network.test.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
+import {
+  InvalidParamsError,
+  UserRejectedRequestError,
+} from '@metamask/snaps-sdk';
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.test.ts:4]:
```suggestion
```

File: packages/starknet-snap/src/rpcs/switch-network.test.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        value: {
+          value: requestNetwork.chainId,
+          markdown: false,
+          type: 'text',
+        },
+      },
+    ]);
+  });
+
+  it('throws `UserRejectedRequestError` if user denied the operation', async () => {
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.test.ts:149]:
```suggestion
  it('throws `UserRejectedOpError` if user denied the operation', async () => {
```

File: packages/starknet-snap/src/rpcs/switch-network.test.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    mockNetworkStateManager({
+      currentNetwork,
+      network: requestNetwork,
+    });
+    const { confirmDialogSpy } = prepareConfirmDialog();
+    confirmDialogSpy.mockResolvedValue(false);
+    const request = createRequestParam(requestNetwork.chainId, true);
+
+    await expect(switchNetwork.execute(request)).rejects.toThrow(
+      UserRejectedRequestError,
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.test.ts:161]:
```suggestion
      UserRejectedOpError,
```

File: packages/starknet-snap/src/rpcs/switch-network.test.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+    await expect(switchNetwork.execute(request)).rejects.toThrow(
+      `Network not supported`,
+    );
+  });
+
+  it('throws `InvalidParamsError` when request parameter is not correct', async () => {
+    await expect(
+      switchNetwork.execute({} as unknown as SwitchNetworkParams),
+    ).rejects.toThrow(InvalidParamsError);
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.test.ts:184]:
```suggestion
    ).rejects.toThrow(InvalidRequestParamsError);
```

File: packages/starknet-snap/src/rpcs/switch-network.ts [hunk 6]
```diff
@@ -0,0 +1,131 @@
+import type { Component } from '@metamask/snaps-sdk';
+import {
+  divider,
+  heading,
+  row,
+  text,
+  UserRejectedRequestError,
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.ts:7]:
```suggestion
```

File: packages/starknet-snap/src/rpcs/switch-network.ts [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import { assign, boolean } from 'superstruct';
+
+import { NetworkStateManager } from '../state/network-state-manager';
+import {
+  confirmDialog,
+  AuthorizableStruct,
+  BaseRequestStruct,
+  RpcController,
+} from '../utils';
+
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.ts:19]:
```suggestion
import { UserRejectedOpError, InvalidNetworkError } from '../utils/exceptions';

```

File: packages/starknet-snap/src/rpcs/switch-network.ts [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        throw new Error(`Network not supported`);
+      }
+
+      if (
+        // Get Starknet expected show the confirm dialog, while the companion doesnt needed,
+        // therefore, `enableAuthorize` is to enable/disable the confirmation
+        enableAuthorize &&
+        !(await this.getSwitchNetworkConsensus(network.name, network.chainId))
+      ) {
+        throw new UserRejectedRequestError() as unknown as Error;
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.ts:92]:
```suggestion
        throw new UserRejectedOpError() as unknown as Error;
```

File: packages/starknet-snap/src/rpcs/switch-network.ts [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+      const network = await networkStateMgr.getNetwork(
+        {
+          chainId,
+        },
+        state,
+      );
+
+      // if the network is not in the list of networks that we support, we throw an error
+      if (!network) {
+        throw new Error(`Network not supported`);
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.ts:83]:
```suggestion
        throw new InvalidNetworkError() as unknown as Error;
```

File: packages/starknet-snap/src/rpcs/switch-network.test.ts [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    // Mock the network state manager to return null network
+    // even if the request chain id is not block by the superstruct
+    mockNetworkStateManager({
+      currentNetwork,
+      network: null,
+    });
+    const request = createRequestParam(requestNetwork.chainId);
+
+    await expect(switchNetwork.execute(request)).rejects.toThrow(
+      `Network not supported`,
```
Comment (khanti42) [inline packages/starknet-snap/src/rpcs/switch-network.test.ts:177]:
```suggestion
      InvalidNetworkError
```

**Review by khanti42 — CHANGES_REQUESTED** [review]
Add suggestion to make it compatible with new error handling from #374

File: packages/starknet-snap/src/rpcs/switch-network.test.ts (hunk 1 above)
Comment (stanleyyconsensys) [reply packages/starknet-snap/src/rpcs/switch-network.test.ts:14]: thank you, but it was expecting to change after prev PR merge, sorry i should put this PR into draft first

**Review by khanti42 — APPROVED** [review]

**Pull request merged**